### ISSUE TITLE
Race condition between sidecar and store

### ISSUE THREAD

**Issue by hsmade** [improbable-eng/thanos#784]
@sudhi-vm

Versions:
Thanos: 0.2.1
Prometheus: 2.7.0
Using the docker images of both.

Setup:
* 4+4 prometheus+sidecars
* 8 stores
* 8 compactors

Problem:
I notice a lot that the store complains that it can't download the meta.json file for new buckets:

`{"caller":"bucket.go:240","err":"new bucket block: load meta: download meta.json: get file: Resource not found","id":"01D2FRNK33V3K8NQYSBVBM24KB","level":"warn","msg":"loading block failed","ts":"2019-01-31T08:52:58.719816484Z"}`

I think that what's happening here is that the sidecar starts uploading to swift, and the store starts picking it up before the upload is finished. Hence it misses the meta.json file.

**Comment by hsmade** [comment]
Another one that I think is related:
```new bucket block: load index cache: download index file: copy object to file: unexpected EOF```
(Also from the store)

**Comment by bwplotka** [comment]
Yup, sounds like valid bug, although it is nor happening that easily on some object stores. 

Wonder if SWIFT is eventually consistent? 

Anyway this needs to be done: https://github.com/improbable-eng/thanos/blob/master/docs/proposals/approved/201901-read-write-operations-bucket.md

But we might start with simple item from this: Handling partial uploads properly (wait for meta json to appear).

**Comment by homelessnessbo** [comment]
I confirm that the same happens with GCS buckets.

**Comment by homelessnessbo** [comment]
Another note: not a big deal, but if you follow examples from here: https://github.com/improbable-eng/thanos/blob/master/examples/alerts/alerts.md - this bug generates false ThanosStoreBucketOperationsFailed alert.

**Comment by bwplotka** [comment]
Yup, example alerts has to be fixed.

Also not sure what is the action item here. It literally does not matter we just produce warning. 

I assume the action item to not warn if that happens to not confuse users, do you agree?

**Comment by bwplotka** [comment]
Does this title make sense?

**Comment by bwplotka** [comment]
Store in the newest version handle partial uploads correctly so this should be closed :+1:
